### PR TITLE
Validate Resources Against US-Core Profiles

### PR DESCRIPTION
**Pull request by SamuilDIntern**
## Description
FHIR resources can now be validated against US-Core profiles using an external terminology service.

## Related issues
N/A

## Testing
**Test Cases:** 
A patient resource, taken from the US-Core implementation guide, was tested for validation against a US-Core patient profile.
A copy of that patient resource was then modified intentionally to not pass the US-core profile and confirm proper profile validation.

**Manual Testing:**
Tested several resource types (patient, observation, allergy Intolerance, and care plan) to verify profile validation.

## FHIR Team Checklist
- [x] **Update the title** of the PR to be succinct and less than 50 characters
- [x] **Add a milestone** to the PR for the sprint that it is merged (i.e. add S47)
- [x] Tag the PR with the type of update: **Bug**, **Dependencies**, **Enhancement**, or **New-Feature**
- [ ] Tag the PR with Azure API for FHIR if this will release to the Azure API for FHIR managed service (CosmosDB or common code related to service)
- [x] Tag the PR with Azure Healthcare APIs if this will release to the Azure Healthcare APIs managed service (Sql server or common code related to service)
- [ ] CI is green before merge
- Review [squash-merge requirements](https://github.com/microsoft/fhir-server/blob/master/SquashMergeRequirements.md)

### Semver Change ([docs](https://github.com/microsoft/fhir-server/blob/master/docs/Versioning.md))
Patch|Skip|Feature|Breaking (reason)

### CODE REVIEW

File: src/Microsoft.Health.Fhir.Api/Resources.resx [hunk 1]
```diff
@@ -387,4 +387,7 @@
     <value>The configuration collection reference '{0}' is invalid.</value>
     <comment>{0} is the configuration collection reference</comment>
   </data>
-</root>
+  <data name="ExportNotSupportedInBundle" xml:space="preserve">
```
Comment (feordin) [inline src/Microsoft.Health.Fhir.Api/Resources.resx:390]: Is this Export string part of your change?

File: src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleHandler.cs [hunk 2]
```diff
@@ -424,6 +424,34 @@ private async Task<EntryComponent> ExecuteRequests(Hl7.Fhir.Model.Bundle respons
                     {
                         HttpContext httpContext = request.HttpContext;
 
+                        string path = httpContext.Request.Path.Value;
```
Comment (feordin) [inline src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleHandler.cs:427]: Is this change related to the Terminology service?

File: src/Microsoft.Health.Fhir.Shared.Web/appsettings.json [hunk 3]
```diff
@@ -86,7 +86,13 @@
                 "OperationTimeout": "00:00:30"
             },
             "Validate": {
-                "CacheDurationInSeconds": 14400
+                "CacheDurationInSeconds": 20000,
+                //"ExternalTerminologyServer": "https://localhost:44349/"
```
Comment (feordin) [inline src/Microsoft.Health.Fhir.Shared.Web/appsettings.json:90]: I don't think we want to check in these changes as default settings.  I think as a default, the ExternalTeminologyServer should not be set.  Once you change this, you will likely need to update the setting in your tests to make sure they work.

File: src/Microsoft.Health.Fhir.Shared.Web/appsettings.json [hunk 4]
```diff
@@ -100,7 +106,7 @@
             "CustomAuditHeaderPrefix": "X-MS-AZUREFHIR-AUDIT-"
         },
         "Bundle": {
-            "EntryLimit": 500
+            "EntryLimit": 5000
```
Comment (feordin) [inline src/Microsoft.Health.Fhir.Shared.Web/appsettings.json:109]: We don't want to check in this change.

File: src/Microsoft.Health.Fhir.Tests.Common/TestFiles/Normative/CodeSystem-careplan-category.json [hunk 5]
```diff
@@ -3,27 +3,42 @@
   "id": "careplan-category",
   "text": {
     "status": "generated",
-    "div": "<div xmlns=\"http://www.w3.org/1999/xhtml\"><h2>US Core CarePlan Category Extension Codes</h2><div><p>Set of codes that are needed for implementation of the US-Core profiles. These codes are used as extensions to the FHIR and US Core value sets.</p>\n</div><p>This code system http://hl7.org/fhir/us/core/CodeSystem/careplan-category defines the following codes:</p><table class=\"codes\"><tr><td style=\"white-space:nowrap\"><b>Code</b></td><td><b>Display</b></td><td><b>Definition</b></td></tr><tr><td style=\"white-space:nowrap\">assess-plan<a name=\"careplan-category-assess-plan\"> </a></td><td>Assessment and Plan of Treatment</td><td>The clinical conclusions and assumptions that guide the patient's treatment and the clinical activities formulated for a patient.</td></tr></table></div>"
+    "div": "<div>PlaceHolder</div>"
   },
   "url": "http://hl7.org/fhir/us/core/CodeSystem/careplan-category",
-  "version": "3.0.0",
+  "version": "4.1.0",
```
Comment (feordin) [inline src/Microsoft.Health.Fhir.Tests.Common/TestFiles/Normative/CodeSystem-careplan-category.json:9]: We do have R4 specific folder for test files.  But if this file works well for all versions of FHIR (including STU3), then changing it here in Normative is fine.

File: src/Microsoft.Health.Fhir.Shared.Web/appsettings.json (hunk 3 above)
Comment (SamuilDIntern) [reply src/Microsoft.Health.Fhir.Shared.Web/appsettings.json:90]: The tests use testConfiguration.json settings I believe, so this file should not affect them. Would it be fine to keep the testConfiguration.json changes?

File: src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleHandler.cs (hunk 2 above)
Comment (SamuilDIntern) [reply src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleHandler.cs:427]: This is from a bug-fix I did my first or second week. I will remove it and make a separate PR for it later.

File: src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleHandler.cs [hunk 6]
```diff
@@ -424,6 +424,8 @@ private async Task<EntryComponent> ExecuteRequests(Hl7.Fhir.Model.Bundle respons
...
```
Comment (LTA-Thinking) [inline src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleHandler.cs:427]: nit: one line of change is left in this file.

Comment (SamuilDIntern) [reply src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleHandler.cs:427]: Fixed.

File: src/Microsoft.Health.Fhir.Tests.Common/Microsoft.Health.Fhir.Tests.Common.csproj [hunk 7]
```diff
@@ -217,6 +220,8 @@
     <None Remove="TestFiles\Stu3\PatientWithMinimalData.json" />
     <None Remove="TestFiles\Stu3\ProvenanceHeader.json" />
     <None Remove="TestFiles\Stu3\Sequence.json" />
+    <None Remove="TestFiles\Stu3\Profile-Patient-PassUsCore-Example.json" />
+    <None Remove="TestFiles\Stu3\Profile-Patient-FailUsCore-Example.json" />
```
Comment (LTA-Thinking) [inline src/Microsoft.Health.Fhir.Tests.Common/Microsoft.Health.Fhir.Tests.Common.csproj:224]: These need to be added as EmbeddedResources like the files in Normative.

Comment (SamuilDIntern) [reply src/Microsoft.Health.Fhir.Tests.Common/Microsoft.Health.Fhir.Tests.Common.csproj:224]: Done.

File: src/Microsoft.Health.Fhir.Api/Resources.resx [hunk 8]
```diff
@@ -1,17 +1,17 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <root>
-  <!--
-    Microsoft ResX Schema
-
+  <!--
```
Comment (feordin) [inline src/Microsoft.Health.Fhir.Api/Resources.resx:3]: There are quite a few lines with whitespace changes here, but I wasn't sure if you had any other changes in the file?  Can it be removed from our PR?

File: src/Microsoft.Health.Fhir.R4.Web/Microsoft.Health.Fhir.R4.Web.csproj [hunk 9]
```diff
@@ -5,6 +5,16 @@
     <RootNamespace>Microsoft.Health.Fhir.Web</RootNamespace>
   </PropertyGroup>
 
+  <ItemGroup>
```
Comment (feordin) [inline src/Microsoft.Health.Fhir.R4.Web/Microsoft.Health.Fhir.R4.Web.csproj:8]: Here and in the other .csproj files, the definitions.zip has a line for Remove= and another line for Include=.  That seems odd, perhaps we don't need the Remove=?

File: src/Microsoft.Health.Fhir.Stu3.Web/Microsoft.Health.Fhir.Stu3.Web.csproj [hunk 10]
```diff
@@ -5,6 +5,17 @@
     <RootNamespace>Microsoft.Health.Fhir.Web</RootNamespace>
   </PropertyGroup>
 
+  <ItemGroup>
+    <None Remove="definitions.zip" />
+    <None Remove="definitions2.zip" />
```
Comment (feordin) [inline src/Microsoft.Health.Fhir.Stu3.Web/Microsoft.Health.Fhir.Stu3.Web.csproj:10]: defintions2.zip?

File: src/Microsoft.Health.Fhir.Stu3.Web/Microsoft.Health.Fhir.Stu3.Web.csproj [hunk 11]
```diff
@@ -30,5 +41,11 @@
     <ProjectReference Include="..\Microsoft.Health.TaskManagement\Microsoft.Health.TaskManagement.csproj" />
   </ItemGroup>
 
+  <ItemGroup>
+    <Resource Include="definitions2.zip">
```
Comment (feordin) [inline src/Microsoft.Health.Fhir.Stu3.Web/Microsoft.Health.Fhir.Stu3.Web.csproj:45]: definitions2.zip?

**Review by feordin — CHANGES_REQUESTED** [review]
It is very close!  I think there is just a little clean up as I mentioned in the comments.

File: src/Microsoft.Health.Fhir.R4.Web/Microsoft.Health.Fhir.R4.Web.csproj (hunk 9 above)
Comment (SamuilDIntern) [reply src/Microsoft.Health.Fhir.R4.Web/Microsoft.Health.Fhir.R4.Web.csproj:8]: That was added automatically when setting the definition.zip as "content" type in the file properties.

I will double check if everything works fine without it.

File: src/Microsoft.Health.Fhir.Api/Resources.resx (hunk 8 above)
Comment (SamuilDIntern) [reply src/Microsoft.Health.Fhir.Api/Resources.resx:3]: Yeah these changes happened automatically when I merged with main I believe? No other changes were made to this file.

Comment (SamuilDIntern) [reply src/Microsoft.Health.Fhir.Api/Resources.resx:3]: These occurred automatically when merging with main, all the contents of the file are the same. Some lines got delete and reinserted back in the same spot but still counted as changes for some reason.

**Review by feordin — APPROVED** [review]
